src/libPersistence/Retriever.cpp

```diff
@@ -75,23 +75,32 @@ void Retriever::RetrieveDSBlocks(bool& result) {
   }
 
   if (isDSIncompleted[0] == '1') {
-    LOG_GENERAL(INFO, "Has incompleted DS Block");
-    if (BlockStorage::GetBlockStorage().DeleteDSBlock(blocks.size() - 1)) {
-      BlockStorage::GetBlockStorage().PutMetadata(MetaType::DSINCOMPLETED,
-                                                  {'0'});
+    if (wakeupForUpgrade) {
+      LOG_GENERAL(INFO, "Has incompleted DS Block, remove it");
+      if (BlockStorage::GetBlockStorage().DeleteDSBlock(blocks.size() - 1)) {
+        BlockStorage::GetBlockStorage().PutMetadata(MetaType::DSINCOMPLETED,
+                                                    {'0'});
+      }
+      blocks.pop_back();
+    } else {
+      LOG_GENERAL(INFO, "Has incompleted DS Block, keep it");
     }
-    blocks.pop_back();
-    hasIncompletedDS = true;
   }
 
+  m_mediator.m_blocklinkchain.Reset();
+  uint64_t index = 0;
+
   for (const auto& block : blocks) {
     m_mediator.m_dsBlockChain.AddBlock(*block);
+    m_mediator.m_blocklinkchain.AddBlockLink(
```

Comment:
this code would be wrong if there is an VC/FB block generated.

Reply:
That's keep the logic here, and once your new AddBlockLink function finished, I will help to test the node recovery functionality.